src/Agent/MsiInstaller/InstallerActions/CustomActions.cs

```diff
@@ -571,9 +571,145 @@ private static void CopyFolderContents(string source, string destination)
                 string fileName = Path.GetFileName(sourceFilePath);
                 string destinationPath = destination + fileName;
 
-                File.Copy(sourceFilePath, destinationPath, true);
-                File.Delete(sourceFilePath);
+                SaferFileUtils.FileCopy(sourceFilePath, destinationPath, true);
+                SaferFileUtils.FileDelete(sourceFilePath);
             }
         }
     }
+
+    /// <summary>
+    /// This class is wrapper around the standard File and Directory classes
+    /// that will check the file and directory structure for symlinks and junctions
+    /// before attempting to copy or delete the file or directory. This code uses
+    /// the presence of reparse points to approximate the usage of junctions or
+    /// symlinks. No agent file or directory is expected to have a reparse point
+    /// defined.
+    /// </summary>
+    internal static class SaferFileUtils
+    {
+        internal static void FileCopy(string source, string destination)
+        {
+            if (!FileIsSafeToUse(source))
+            {
+                return;
```

Comment:
Feels like some logging would be helpful for troubleshooting so we know why a particular operation wasn't performed.

Reply:
Sounds like a good idea. I'll need to see if there is already a way for us to log things from within the installer.